pkg/domain/infra/abi/quadlet.go

```diff
@@ -0,0 +1,764 @@
+	// First, validate that the source path exists and is a file
+	stat, err := os.Stat(path)
+	if err != nil {
+		return "", fmt.Errorf("quadlet to install %q does not exist or cannot be read: %w", path, err)
+	}
+	if stat.IsDir() {
+		return "", fmt.Errorf("quadlet to install %q is not a file", path)
+	}
+
+	finalPath := filepath.Join(installDir, filepath.Base(filepath.Clean(path)))
+	if destName != "" {
+		finalPath = filepath.Join(installDir, destName)
+	}
+
+	// Validate extension is valid
+	if isQuadletFile && !systemdquadlet.IsExtSupported(finalPath) {
+		return "", fmt.Errorf("%q is not a supported Quadlet file type", filepath.Ext(finalPath))
+	}
+
+	// Second, validate that the dest path does NOT exist.
+	// TODO: Overwrite flag?
+	if _, err := os.Stat(finalPath); err == nil {
+		return "", fmt.Errorf("a Quadlet with name %s already exists, refusing to overwrite", filepath.Base(finalPath))
+	}
+
+	file, err := os.OpenFile(finalPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0644)
+	if err != nil {
+		return "", fmt.Errorf("unable to open file %s: %v", filepath.Base(finalPath), err)
+	}
+	defer file.Close()
+
+	// Move the file in
+	srcFile, err := os.Open(path)
+	if err != nil {
+		return "", fmt.Errorf("unable to open file %s: %v", path, err)
+	}
+	defer srcFile.Close()
+
+	err = fileutils.ReflinkOrCopy(srcFile, file)
+	if err != nil {
+		return "", fmt.Errorf("unable to copy file from %s to %s: %v", path, finalPath, err)
+	}
+
+	// When we install files using this function, caller of this function can turn off `validateQuadletFile`
+	// when they are installing `non-quadlet` files.
+	if !isQuadletFile {
+		err := appendStringToFile(filepath.Join(installDir, assetFile), filepath.Base(filepath.Clean(path)))
+		if err != nil {
+			return "", fmt.Errorf("error while writing non-quadlet filename: %w", err)
+		}
+	}
+	return finalPath, nil
+}
+
+// appendStringToFile appends the given text to the specified file.
+// If the file does not exist, it will be created with 0644 permissions.
+func appendStringToFile(filePath, text string) error {
+	f, err := os.OpenFile(filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	_, err = f.WriteString(text + "\n")
+	return err
+}
+
+// buildAppMap scans the given directory for files that start with '.'
+// and end with '.app', reads their contents (one filename per line), and
+// returns a map where each filename maps to the .app file that contains it.
+// Also returns a map where each `.app` points to a slice of strings containing
+// all the files in that `.app`.
+func buildAppMap(dir string) (map[string]string, map[string][]string, error) {
+	reverseMap := make(map[string]string)
+	appMap := make(map[string][]string)
+
+	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
```

Comment:
this can already fail with ErrNotExist so this must be ignored here too.
`podman quadlet list` is completely broken right now when the `~/.config/containers/systemd/` doesn't exist because it fails to builds the appMap here